File: saizad-mvvm/src/main/java/com/saizad/mvvm/FCMToken.java
package com.saizad.mvvm;

import android.content.SharedPreferences;

import androidx.annotation.Nullable;

import com.google.gson.Gson;
import com.sa.easyandroidfrom.ObjectUtils;

import io.reactivex.Observable;
import io.reactivex.subjects.BehaviorSubject;

public class FCMToken extends SaizadSharedPreferences {

    private static final String KEY_FCM_TOKEN = "fcm_token";
    private final BehaviorSubject<String> fcmTokenSubject = BehaviorSubject.create();

    public FCMToken(SharedPreferences sharedPreferences, Gson gson) {
        super(sharedPreferences, gson);
    }

    public void putToken(@Nullable String fcmToken){
        putValue(KEY_FCM_TOKEN, fcmToken);
        fcmTokenSubject.onNext(ObjectUtils.coalesce(fcmToken, ""));
    }

    public String getFcmToken(){
        return getValue(KEY_FCM_TOKEN);
    }

    public Observable<String> observeFcmToken(){
        return fcmTokenSubject.filter(s -> !s.isEmpty());
    }

    public Observable<String> observeFcmTokenRemoved(){
        return fcmTokenSubject.filter(String::isEmpty);
    }
}

File: saizad-mvvm/src/main/java/com/saizad/mvvm/FCMToken.kt
package com.saizad.mvvm

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.asFlow
import com.shopify.livedataktx.filter
import com.shopify.livedataktx.map
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.drop
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.launch
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class FCMToken @Inject constructor(private val dataStoreWrapper: DataStoreWrapper) {

    val tokenLiveData = MutableLiveData<String?>()
    private var token: String? = null

    companion object {
        private const val KEY_FCM_TOKEN = "fcm_token"
    }

    init {

        GlobalScope.launch {
            tokenLiveData
                .asFlow()
                .drop(1)
                .collect {
                    token = it
                    if (it == null) {
                        dataStoreWrapper.remove(KEY_FCM_TOKEN)
                    } else {
                        dataStoreWrapper.putObject(KEY_FCM_TOKEN, it as Any)
                    }
                }
        }

        GlobalScope.launch {
            tokenLiveData.postValue(
                dataStoreWrapper.getValue(KEY_FCM_TOKEN).firstOrNull()
            )
        }

    }

    fun putToken(fcmToken: String?) {
        tokenLiveData.postValue(fcmToken)
    }

    val fcmToken: String?
        get() = token

    fun observeFcmToken(): LiveData<String> {
        return tokenLiveData.filter { !it.isNullOrEmpty() }.map { it!! }
    }

    fun observeFcmTokenRemoved(): LiveData<Void?> {
        return tokenLiveData.filter { it.isNullOrEmpty() }.map { null }
    }


}